stop injecting the container when possible

xabbuh opened the pull request:
Injecting the container in your services is not a good practice. After this, there are only three classes left that rely on the container being injected: The `RestRouteLoader` class needs it to resolve controllers registered as services, the `ParamFetcher` class injects the container into container aware `ParamInterface` implementations, and the parent class of the `ViewResponseListener` uses the container.

xabbuh commented:
This is ready to be reviewed.

Ener-Getick commented on EventListener/ParamFetcherListener.php:
```diff
@@ -24,18 +24,18 @@
      *
-     * @param ContainerInterface $container
-     * @param bool               $setParamsAsAttributes
+     * @param ParamFetcher $paramFetcher
```
There is a ``ParamFetcherInterface``. I think it would be better to use it.

stof replied:
indeed

Ener-Getick commented:
+1

Ener-Getick commented on View/ViewHandler.php:
```diff
@@ -315,7 +340,7 @@ protected function getTemplating()
     {
         if (null === $request) {
-            $request = $this->container->get('request_stack')->getCurrentRequest();
+            $request = $this->requestStack->getCurrentRequest();
```
``RequestStack::getCurrentRequest()`` can also be null, maybe add a check ?

stof replied:
this is indeed needed to avoid fatal errors in weird cases

xabbuh replied:
That's true but we need to fix this in previous versions.

lsmith77 commented:
IIRC we injected the container in the ExceptionController and ViewHandler mostly because depending on the code paths, not all dependencies were needed. These days there are of course proxy services which could be used here as well.

lsmith77 commented:
@xabbuh so how to best address the performance concerns in regards to optional dependencies ..? I guess our options are 1) leave as is 2) your proposal 3) use proxy service. I guess the proper way to do this would be to try and benchmark this ..

we can of course look for some blog posts (f.e. https://blog.cloudtroopers.com/symfony-2-lazy-loading-services-and-performance-analysis) and try and extrapolate the relevance. maybe we can also look a bit more closely which services are actually expensive to create and which are actually fairly cheap.

xabbuh commented:
@lsmith77 Do you remember which of the injected services were expected to be slow to initialize? On a quick look it doesn't seem to me that it's actually expensive to construct them.

lsmith77 commented on Decoder/DecoderProvider.php:
```diff
@@ -11,21 +11,19 @@
  * @author Igor Wiedler <igor@wiedler.ch>
  */
-class ContainerDecoderProvider extends ContainerAware implements DecoderProviderInterface
+class DecoderProvider implements DecoderProviderInterface
```
this is a good example of a service where I think for performance reasons it would make sense to keep the container injection

xabbuh replied:
Agreed, I reverted this change.

xabbuh replied:
Well, maybe we should inject the container in the constructor as it is required anyway. What do you think?

lsmith77 commented on Controller/ExceptionController.php:
```diff
@@ -26,8 +27,34 @@
+        ExceptionWrapperHandlerInterface $exceptionWrapperHandler,
+        FormatNegotiator $formatNegotiator,
+        ViewHandlerInterface $viewHandler,
+        EngineInterface $templating,
```
I would make this a lazy service, since it might never be needed for an API only setup

stof replied:
making the service lazy would be hard (as you don't own the service, and it would add overhead to anyone using it all the time because of the proxy wrapper).

However, there is something else needed here: 

- it should be possible to build this service when there is no templating system in the project (assuming you don't use the HTML format)
- I would suggest using Twig directly instead of the templating component wrapping twig, as you only render Twig templates anyway

lsmith77 replied:
good point .. this means we need to dynamically determine in the extension if to inject the service .. or alternatively we can use the syntax to optionally inject the twig service only if its available

xabbuh replied:
Good idea, though which interface should we rely on then?

lsmith77 replied:
BTW same applies to the ViewHandler

lsmith77 replied:
here is an attempt to solve the issue:
https://github.com/FriendsOfSymfony/FOSRestBundle/commit/93758654c0ac2ed29576c0dcbe7d42122d71bf37

maybe we should do something similar for the ViewHandler?

lsmith77 commented on View/ViewHandler.php:
```diff
@@ -104,17 +106,33 @@ class ViewHandler extends ContainerAware implements ConfigurableViewHandlerInter
     public function __construct(
+        RouterInterface $router,
+        $serializer,
+        EngineInterface $templating,
```
I would make this a lazy service, since it might never be needed for an API only setup

lsmith77 commented on View/ViewHandler.php:
```diff
@@ -104,17 +106,33 @@ class ViewHandler extends ContainerAware implements ConfigurableViewHandlerInter
      */
     public function __construct(
+        RouterInterface $router,
+        $serializer,
```
I would make this a lazy service, since it might never be needed for an HTML request

xabbuh replied:
Not sure if that works `fos_rest.serializer` only is an alias. What about still injecting the container and retrieving the serializer from it? We then just have to think about what we do at runtime if the service doesn't exist.

lsmith77 replied:
imho its also worth consider using setter injection for those "optional" or non strongly typed dependencies and then using an internal getter that handles the case of a missing dependency

lsmith77 replied:
BTW for the serializer we could also consider creating an adapter so that for the ViewHandler we can assume a specific interface

lsmith77 commented on View/ViewHandler.php:
```diff
@@ -104,17 +106,33 @@ class ViewHandler extends ContainerAware implements ConfigurableViewHandlerInter
+        $serializer,
+        EngineInterface $templating,
+        RequestStack $requestStack,
+        ExceptionWrapperHandlerInterface $exceptionWrapperHandler,
```
not sure if this should also be a lazy service .. I guess its fairly cheap ..

lsmith77 commented:
@xabbuh I commented inline

lsmith77 commented:
ping

lsmith77 commented:
need a rebase

xabbuh commented on Tests/EventListener/ParamFetcherListenerTest.php:
```diff
@@ -79,6 +74,7 @@ public function testSettingParamFetcherOnRequest()
 
     public function testParamFetcherOnRequestNoZone()
     {
+        $this->markTestSkipped();
```
I am wondering how the test passed before as the event does not return a callable controller.

xabbuh replied:
Got it, I simply deleted too much from the `ParamFetcherListener` class.

lsmith77 commented:
please rebase again to get the 3.0 build issue fixed

lsmith77 commented:
@xabbuh FYI I have you write permissions on this repo .. as you know its still best to work through PRs and branches but this makes it easier to collaborate

xabbuh commented:
I did two more things here:

* Removed the `ContainerAwareInterface` usages where we already always inject the container through a setter. The only places left being "container aware" should now be the request params.
* Introduced adapter for the JMS serializer and Symfony's core serializer. Now we can use lazy services and type hints here. I just wonder if we now also could get rid the of context adapters.

Ener-Getick commented on Context/Adapter/JMSContextAdapter.php:
```diff
@@ -74,7 +76,7 @@ public function convertDeserializationContext(ContextInterface $context)
     public function supportsSerialization(ContextInterface $context)
     {
-        return $this->serializer instanceof JMSSerializerInterface;
+        return $this->serializer instanceof JmsSerializerAdapter;
```
May be better to use the same vendor syntax as in [``jms/serializer``](https://github.com/schmittjoh/serializer/tree/master/src/JMS/Serializer):
``JMSSerializerAdapter``.

Ener-Getick commented:
:+1: with this change, we could move the ``Context`` transformation in ``Serializer``.

lsmith77 commented:
```
1) FOS\RestBundle\Tests\Functional\RequestBodyParamConverterTest::testRequestBodyIsDeserialized
Argument 1 passed to FOS\RestBundle\Tests\Functional\Bundle\TestBundle\Controller\RequestBodyParamConverterController::putPostAction() must be an instance of FOS\RestBundle\Tests\Functional\Bundle\TestBundle\Controller\Post, null given
```

lsmith77 commented on DependencyInjection/Compiler/SerializerConfigurationPass.php:
```diff
@@ -33,13 +33,26 @@ public function process(ContainerBuilder $container)
+            } elseif (is_subclass_of($class, 'JMS\Serializer\SerializerInterface')) {
+                $container->setAlias('fos_rest.serializer', 'fos_rest.serializer.jms');
+            } else {
+                throw new \LogicException(sprintf('Cannot handle serializer of type "%s"', $class));
```
I guess this is one of the draw backs of this .. ie. we now hardcode exactly which Serializer implementations we support

xabbuh replied:
Well, the idea is to add a config option in the future that would make it possible to inject/create your own serializer service. The compiler pass then would exit earlier [here](https://github.com/xabbuh/FOSRestBundle/blob/service-container/DependencyInjection/Compiler/SerializerConfigurationPass.php#L27-L29) without throwing an exception.

stof replied:
well, to support everything, the compiler pass should also accept configuring a service implementing directly the FOSRestBundle interface

xabbuh replied:
I agree and that's what I mean with the early exit in the compiler pass when the `fos_rest.serializer` service is already set (which we may ease later by adding a config option for it). Or do you mean something else?

lsmith77 commented:
I guess we should not overburden this PR .. so imho lets get this wrapped up and then do more things in other PRs.

I still think that we should also do something like https://github.com/FriendsOfSymfony/FOSRestBundle/commit/93758654c0ac2ed29576c0dcbe7d42122d71bf37 with all the places where we inject the templating engine.

xabbuh commented:
@lsmith77 I agree with you (93758654c0ac2ed29576c0dcbe7d42122d71bf37 actually is the thing I talked about I would like to include). On the other hand @Ener-Getick and me were talking about backporting certain features to a (potential) 1.8 release to be able to create true BC layers and to provide a smooth upgrade path (see https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1213#discussion_r45556284). What do you think about this?

lsmith77 commented:
I must admit .. I personally do not think such back ports are worth the effort

Ener-Getick commented:
@xabbuh I merged the master branch and fixed the tests in [this branch](https://github.com/FriendsOfSymfony/FOSRestBundle/tree/CONTAINER) if it helps you ;)

stof commented:
@lsmith77 if you don't provide a smooth upgrade path from 1.x to 2.x, it means you will need to provide a LTS for 1.x as people would not be able to upgrade all projects.
And it will also make it much harder for bundles depending on FOSRestBundle as they would have to either support only one version (splitting the community between people able to upgrade or now) or make crazy stuff to support both without a simple way to write compatible code.

Note that BC layers are mostly necessary for changes impacting the places using the bundle feature, for instance the annotations.
Bundles extending internal parts of the bundle may be left in front of the above difficulties if you want (but a smooth upgrade path for everyone would be easier)

Ener-Getick commented on Resources/config/serializer.xml:
```diff
@@ -0,0 +1,15 @@
+
+    <services>
+        <service id="fos_rest.serializer.jms" class="FOS\RestBundle\Serializer\JmsSerializerAdapter" public="false" lazy="true">
+            <argument type="service" id="jms_serializer.serializer" />
```
This will throw an error if the service doesn't exist, right ?

xabbuh replied:
I think in this case the service should already be removed as it is an unused private service then (IIRC we have functional tests that cover this case).

xabbuh commented:
@lsmith77 I agree with you that we cannot do that for each and every part. But I also agree with @stof that we should provide a clear and smooth upgrade path to ease the life for other bundles that depend on FOSRestBundle.

I believe defining first which parts of the bundle are internal and which expose our public API would be a big step in the right direction (everything not being part of the public API would not necessarily have to provide a smooth upgrade path). For that reason I suggest that we start marking all classes that are not meant to be used by consumer code as internal (see https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1213#issuecomment-158617633). What do you think?

xabbuh commented:
@Ener-Getick Ah that did the trick. What a stupid mistake. Thank you very much!

stof commented on Serializer/Serializer.php:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * @param mixed  $data
+     * @param string $format
+     * @param mixed  $context
```
this should probably be more specific than ``mixed``, abstracting the differences

Ener-Getick replied:
As @xabbuh said [here](https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1171#issuecomment-158617224) IMO we should the context adapters behavior in the serializer adapters and change this type hint to ``ContextInterface``.
May be better to do this later in another PR as this one becomes really big.

stof commented on EventListener/ViewResponseListener.php:
```diff
@@ -27,16 +28,22 @@
  */
 class ViewResponseListener extends TemplateListener
```
this should not extend the TemplateListener anymore but just run earlier (and so no need for the container)

Ener-Getick replied:
This is causing a bug see https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1301.
Should we revert this change or force the user to always enable the ``TemplateListener`` ?

Ener-Getick replied:
May be better to just revert this change or the ``ViewResponseListener`` will be dependant of the ``SensioExtraFrameworkBundle``

stof commented on View/ViewHandler.php:
```diff
@@ -107,17 +107,33 @@ class ViewHandler implements ConfigurableViewHandlerInterface, ContainerAwareInt
+     * @param string                           $defaultEngine           default engine (twig, php ..)
      */
     public function __construct(
+        RouterInterface $router,
```
We should only inject a UrlGeneratorInterface here (changing the ``getRouter``)

stof commented on View/ViewHandler.php:
```diff
@@ -107,17 +107,33 @@ class ViewHandler implements ConfigurableViewHandlerInterface, ContainerAwareInt
     public function __construct(
+        RouterInterface $router,
+        Serializer $serializer,
+        EngineInterface $templating,
```
We should allow injecting Twig_Environment directly too.

stof replied:
and should be optional

Ener-Getick commented on Resources/config/serializer.xml:
```diff
@@ -0,0 +1,15 @@
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="fos_rest.serializer.jms" class="FOS\RestBundle\Serializer\JmsSerializerAdapter" public="false" lazy="true">
```
Must be changed to ``FOS\RestBundle\Serializer\JMSSerializerAdapter`` ;-)

Ener-Getick commented:
@xabbuh do you need help on this PR ?

xabbuh commented:
@Ener-Getick Rebasing was a nightmare. ;) I need to resolve some conflicts and move the serialization API to a new pull request that will go to the `1.8` branch.

Ener-Getick commented:
Already done in https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1252 :P

xabbuh commented:
@Ener-Getick Ah right, I didn't think about that. I think we still need to simplify that a bit, but I will comment directly on your PR.

Let's see if all tests still pass and then I need to check all comments again and see if I really addressed all of your comments.

Ener-Getick commented on Decoder/ContainerDecoderProvider.php:
```diff
@@ -11,27 +11,27 @@
  * @author Igor Wiedler <igor@wiedler.ch>
  */
-class ContainerDecoderProvider implements DecoderProviderInterface, ContainerAwareInterface
+class ContainerDecoderProvider implements DecoderProviderInterface
```
Not sure this name is still relevant

Ener-Getick commented:
Ready to merge for me.
Do you want to include other changes @xabbuh?

Ener-Getick commented:
Status: reviewed

lsmith77 commented:
needs a rebase

Ener-Getick commented:
See #1286
